<launch>
  <arg name="launch_main" default="true" />
  <arg name="json" />

  <!-- shelf-->
  <arg name="UPPER_SHELF" value="$(find jsk_apc2016_common)/config/bin_upper_shelf.yaml"/>
  <arg name="LOWER_SHELF" value="$(find jsk_apc2016_common)/config/bin_lower_shelf.yaml"/>

  <!-- Publish jsk_apc2016_common/BinInfo from json -->
  <node name="publish_bin_bbox"
        pkg="jsk_apc2016_common" type="publish_bin_bbox.py"
        output="screen">
    <rosparam file="$(arg UPPER_SHELF)" command="load" ns="upper_shelf"/>
    <rosparam file="$(arg LOWER_SHELF)" command="load" ns="lower_shelf"/>
  </node>

  <node pkg="jsk_2016_01_baxter_apc" type="stow_work_order_server.py" name="work_order" respawn="true" output="screen" clear_params="true">
    <rosparam subst_value="true">
        json: $(arg json)
        limit_volume: 5000
        volume_first:
          - dasani_water_bottle
          - dr_browns_bottle_brush
          - folgers_classic_roast_coffee
          - hanes_tube_socks
          - kleenex_tissue_box
          - peva_shower_curtain_liner
          - up_glucose_bottle
        black_list: []
    </rosparam>
  </node>

  <!-- parameter -->
  <param name="/left_hand/state" value="" />
  <param name="/left_hand/target_bin" value="" />
  <param name="/left_hand/target_object" value="" />
  <param name="/left_hand/update_json" value="false" />
  <param name="/right_hand/state" value="" />
  <param name="/right_hand/target_bin" value="" />
  <param name="/right_hand/target_object" value="" />
  <param name="/right_hand/update_json" value="false" />

  <node pkg="jsk_rqt_plugins" type="rqt_yn_btn" name="rqt_yn_btn" />

  <node pkg="jsk_2016_01_baxter_apc" type="output_json_stow.py" name="output_json_stow" respawn="true">
    <rosparam subst_value="true">
        json: $(arg json)
    </rosparam>
  </node>

  <!-- object verification -->
  <group ns="left_hand">
    <node name="object_verification"
          pkg="jsk_2016_01_baxter_apc" type="apply_tote_contents_hint.py"
          respawn="true">
      <remap from="~input" to="vgg16_object_recognition/output" />
      <param name="json" value="$(arg json)" />
    </node>
  </group>
  <group ns="right_hand">
    <node name="object_verification"
          pkg="jsk_2016_01_baxter_apc" type="apply_tote_contents_hint.py"
          respawn="true">
      <remap from="~input" to="vgg16_object_recognition/output" />
      <param name="json" value="$(arg json)" />
    </node>
  </group>

  <!-- robot main -->
  <node if="$(arg launch_main)" pkg="jsk_2016_01_baxter_apc" type="main-stow.l" name="right_hand"
    args="&quot;(progn (apc-stow-init :rarm-controller) (apc-stow-mainloop :rarm))&quot;"
    respawn="true" output="screen">
    <env name="DISPLAY" value="" />
  </node>
  <node if="$(arg launch_main)" pkg="jsk_2016_01_baxter_apc" type="main-stow.l" name="left_hand"
    args="&quot;(progn (apc-stow-init :larm-controller) (apc-stow-mainloop :larm))&quot;"
    respawn="true" output="screen">
    <env name="DISPLAY" value="" />
  </node> 

  <!-- visualization -->
  <node name="visualize_json"
        pkg="jsk_apc2016_common" type="visualize_stow_json.py"
        args="$(arg json)" />
  <!--
  <node name="visualize_on_rviz"
        pkg="jsk_2015_05_baxter_apc" type="visualize_on_rviz.py">
    <param name="json" value="$(arg json)" />
  </node>
  -->
  <node name="rosconsole_overlay_text"
        pkg="jsk_rviz_plugins" type="rosconsole_overlay_text.py">
    <rosparam>
      nodes:
        - /left_hand
        - /right_hand
    </rosparam>
  </node>

  <!-- set rosparam for fcn -->
  <node name="set_tote_contents_param"
        pkg="jsk_2016_01_baxter_apc" type="set_tote_contents_param.py"
        args="$(arg json)">
    <remap from="~tote_contents" to="/apc2016_tote_contents" />
  </node>

</launch>
